File: runtime-environment/process-communication/cpp_comm_interface/src/CommInterface.cpp
/**
 * @file CommInterface.cpp
 * @brief Contains the Publisher and Subscriber classes for communication between processes using ZeroMQ.
 * 
 * Provides a C++ implementation of the ZeroMQ publisher-subscriber communication pattern. 
 * The Publisher class sends messages to one or more subscribers using a publisher socket, and the 
 * Subscriber class receives messages from a publisher using a subscriber socket.
 * 
 * @author Onur Ulusoy
 * @date 22.04.2023
 * @license MIT
 */

#include "CommInterface.hpp"

Publisher::Publisher(const std::string& localAddress, const std::string& processName)
    : context(1), socket(context, ZMQ_PUB), processName(processName) {
    std::string port = localAddress.substr(localAddress.find_last_of(':') + 1);
    std::string endpoint = "tcp://*:" + port;
    socket.bind(endpoint);
}

void Publisher::publish(const std::string& message) {
    std::string messageWithProcessName = processName + ": " + message;
    msgpack::sbuffer buffer;
    msgpack::pack(buffer, messageWithProcessName);
    zmq::message_t packedMessage(buffer.data(), buffer.size());
    socket.send(packedMessage, zmq::send_flags::none);
    std::cout << "Sent: " << messageWithProcessName << std::endl;
}

void Publisher::close() {
    socket.close();
    context.close();
}

Subscriber::Subscriber(const std::string& localAddress)
    : context(1), socket(context, ZMQ_SUB) {
    socket.connect(localAddress);
    socket.set(zmq::sockopt::subscribe, "");  // Updated function call
}

std::string Subscriber::receive(int timeout) {
    zmq::message_t packedMessage;
    socket.set(zmq::sockopt::rcvtimeo, timeout);

    auto result = socket.recv(packedMessage, zmq::recv_flags::none);  // Handle the return value
    if (result) {
        msgpack::object_handle oh = msgpack::unpack(static_cast<const char*>(packedMessage.data()), packedMessage.size());
        msgpack::object deserialized = oh.get();
        std::string message;
        deserialized.convert(message);
        std::cout << "Received: " << message << std::endl;

        size_t delimiterPos = message.find(':');
        if (delimiterPos != std::string::npos) {
            message = message.substr(delimiterPos + 1);
            message = message.substr(message.find_first_not_of(' '));
        }
        return message;
    } else if (timeout > 0 && zmq_errno() == EAGAIN) {
        std::cout << "Timeout occurred while waiting for a message." << std::endl;
        return "";
    }

    return "";
}

void Subscriber::close() {
    socket.close();
    context.close();
}

File: runtime-environment/process-communication/comm_interface.py
"""
@file comm_interface.py
@brief Contains the Publisher and Subscriber classes for communication between processes using ZeroMQ.

This module provides a Python implementation of the ZeroMQ publisher-subscriber communication pattern. 
The Publisher class sends messages to one or more subscribers using a publisher socket, and the 
Subscriber class receives messages from a publisher using a subscriber socket.

@author Onur Ulusoy
@date 22.04.2023
@license MIT
"""

import zmq
import msgpack

class Publisher:
    """
    Class representing a ZeroMQ publisher.
    @details This class allows for sending messages to a ZeroMQ subscriber using a publisher socket.
    """
    def __init__(self, local_address, process_name):
        """
        Constructor for Publisher object.
        
        @param local_address The local address to bind the socket to.
        @param process_name The name of the process that is publishing the message.
        """
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.PUB)
        self.socket.bind(f"tcp://*:{local_address.split(':')[-1]}")  # Bind using the wildcard
        self.process_name = process_name

    def publish(self, message):
        """
        Sends a message using the Publisher object's socket.
        
        @param message The message to be sent.
        """
        message_with_process_name = f"{self.process_name}: {message}"
        packed_message = msgpack.packb(message_with_process_name)
        self.socket.send(packed_message)
        print(f"Sent: {message_with_process_name}")
    
    def close(self):
        """
        @brief Close the publisher socket and terminate the ZeroMQ context.

        @details This method closes the socket associated with the Publisher object and
        terminates the ZeroMQ context, which releases any resources being used by the socket
        and context.
        """
        self.socket.close()
        self.context.term()

class Subscriber:
    """
    Class representing a ZeroMQ subscriber.
    @details This class allows for receiving messages from a ZeroMQ publisher using a subscriber socket.
    """
    def __init__(self, local_address):
        """
        Constructor for Subscriber object.
        
        @param local_address The local address to connect the socket to.
        """
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.SUB)
        self.socket.connect(local_address)
        self.socket.setsockopt_string(zmq.SUBSCRIBE, "")

    def receive(self, timeout=None):
        """
        Receives a message using the Subscriber object's socket.
        
        @details The received message is deserialized using the msgpack library and printed to the console.
        @param timeout: The maximum time to wait for a message, in milliseconds. None (default) means wait indefinitely.
        """
        self.socket.setsockopt(zmq.RCVTIMEO, timeout if timeout is not None else -1)
        try:
            packed_message = self.socket.recv()
            message = msgpack.unpackb(packed_message, raw=False)
            print(f"Received: {message}")
            
            message = message.split(':', 1)[-1].strip()
            return message
        
        except zmq.error.Again:
            print("Timeout occurred while waiting for a message.")
            return None
    
    def close(self):
        """
        @brief Close the subscriber socket and terminate the ZeroMQ context.

        @details This method closes the socket associated with the Subscriber object and
        terminates the ZeroMQ context, which releases any resources being used by the socket
        and context.
        """
        self.socket.close()
        self.context.term()


